guide: update COW link note (#4144)
iterative/dvc.org: content/docs/user-guide/data-management/large-dataset-optimization.md

---

<admon>
   
\* ([copy-on-write](https://en.wikipedia.org/wiki/Copy-on-write)) links or
**reflinks** are a type of file linking available in UNIX-style file systems.

> Unlike hardlinks or symlinks, these file entries turn into a plain copy when

Reviewer: this is not true, file doesn't become a copy -- only certain part of the file becomes a copy 

what was the reason to touch this?

Author: This was [contributed](https://github.com/iterative/dvc.org/pull/4127#pullrequestreview-1190702248) by someone external who identified that the previous text was incorrect. I'm just applying it here as well (this note exists in a couple places).

Reviewer: got it, yes, text is still misleading a bit ...

Author: I generalized the statement. PTAL

> Unlike hard links or symlinks, these file entries are only loosely connected to

Reviewer: Let's not use terms like "loose". Let's just explicitly explain if it's needed ... that only small part is being copied when/if edit is happening

Author: TBH I don't know the exact mechanics and I suspect they may change from file system to file system. In general from what I can see out there, CoW means literally to create a copy of the resource (usually an entire file, I assume) when said resource is modified.

I'm not sure why we want to to explain this very low-level operation in so much detail, though. Can we keep it general like just remove the sentence about "Unlike hard links or symlinks..." ?

Author: Note I requested a review form the core team from the beginning but they didn't seem to notice. Should we ping someone specific? I have the impression that @efiop will know this one 🙂

Reviewer 2: It's not necessarily the entire file that gets copied. reflinks work at the filesystem block level, not the file level. So on a write, only the particular block that would be modified is copied into the new block.

Also, copy-on-write/reflinks aren't unix specific. ReFS supports copy-on-write operations on windows (but DVC does not support reflinking on windows)

Reviewer 2: A reflink is essentially a list of references to filesystem blocks. So when you create a reflink, it will contain the same list of blocks as the file you are linking to.

Copy-on-write filesystem blocks are never modified in-place. On a write, the block is copied into a new one, and then write operation is done in that new block. This way, any existing files/reflinks that pointed to the original block remain unchanged (because the original block is also unchanged). Only specific file the user is writing to will have its list of references updated (w/a reference to the new/modified block replacing the original one)

Reviewer 2: From a docs standpoint, I'm not sure we need to get into details of how each link type actually works?

The main point to get across should be that for cases where reflinks are available (cache is on the same filesystem as the workspace, and the filesystem supports reflinks), reflinks are the best option in terms of both speed and storage usage, and do not have any real drawbacks.

Author: > reflinks work at the filesystem block level

Thanks for the info, makes sense. I also don't think we need to go into this kind of detail in our docs. I'll remove that part of the explanation for now...

> copy-on-write/reflinks aren't unix specific

Good point. Updating...

> for cases where reflinks are available... reflinks are the best option.

Yep. This should already be noted earlier in these and in other docs. We're just trying to update these admonitions about what CoW is.

> \* ([copy-on-write](https://en.wikipedia.org/wiki/Copy-on-write)) links or
> **reflinks** are a type of file linking available in UNIX-style file systems.
> Unlike hard links or symlinks, these file entries are only loosely connected to
> the original content. This means that editing reflinks is always safe, as the
> original <abbr>cached</abbr> data will remain unchanged.

Author: ```suggestion
\* ([copy-on-write](https://en.wikipedia.org/wiki/Copy-on-write)) links or
**reflinks** are a type of file linking available in modern file systems. Unlike
hard links or symlinks, editing reflinks is always safe, as the original
<abbr>cached</abbr> data will remain unchanged.
```
